File: week_01/Python_code/subsets.py
class Solution:
    """
    @param S: The set of numbers.
    @return: A list of lists. See example.
    """
    def subsets(self, S):
        if not S:
            return []
        result = []
        S.sort()
        self.helper(S, [], result, 0)
        return result

    def helper(self, S, lis, result, index):
        if index == len(S):
            result.append(lis)
            return
        self.helper(S, lis, result, index + 1)
        self.helper(S, lis + [S[index]], result, index + 1)

if __name__ == '__main__':
    test = Solution()
    S = [1,2,3]
    print test.subsets(S)

File: NineChapters/01.strStr_CodingStyle/Java_Code/Subsets.java
import java.util.ArrayList;
import java.util.Collections;

public class Subsets {
    /**
     * @param S: A set of numbers.
     * @return: A list of lists. All valid subsets.
     */
    public ArrayList<ArrayList<Integer>> subsets(ArrayList<Integer> S) {
        ArrayList<ArrayList<Integer>> result = new ArrayList<ArrayList<Integer>>();
        if (S == null || S.size() == 0) {
            return result;
        }
        Collections.sort(S);
        ArrayList<Integer> list = new ArrayList<Integer>();
        subsetsHelper(S, result, list, 0);
        return result;
    }

    private void subsetsHelper(ArrayList<Integer> S, ArrayList<ArrayList<Integer>> result,
        ArrayList<Integer> list, int index) {
        result.add(new ArrayList<Integer>(list));
        for (int i = index; i < S.size(); i++) {
            list.add(S.get(i));
            subsetsHelper(S, result, list, i + 1);
            list.remove(list.size() - 1);
        }
    }

    public static void main(String[] args) {
        Subsets test = new Subsets();
        ArrayList<Integer> S = new ArrayList<Integer>();
        S.add(1);
        S.add(2);
        S.add(3);
        System.out.println(test.subsets(S));
    }
}
